security-analysis/security-analysis-api/src/main/java/com/powsybl/security/SecurityAnalysisImpl.java

```diff
@@ -88,48 +89,76 @@ public boolean removeInterceptor(SecurityAnalysisInterceptor interceptor) {
-
-                                        return null;
-                                    }, computationManager.getExecutor());
+                        int workerCount = Math.min(computationManager.getResourcesStatus().getAvailableCores(), contingencies.size());
+                        BlockingQueue<Integer> queue = new ArrayBlockingQueue<>(workerCount);
+                        for (int i = 0; i < workerCount; i++) {
+                            String postContStateId = hash + "_" + i;
+                            network.getVariantManager().cloneVariant(VariantManagerConstants.INITIAL_VARIANT_ID, postContStateId);
+                            queue.add(i); //Will always work because we are initializing the queue and the initial capacity is workerCount
                         }
+                        // use completedFuture(null).thenCompose so that more of the
+                        // execution is handled by the whenComplete block removing the
+                        // variants.
+                        future = CompletableFuture.completedFuture(null).thenComposeAsync(aVoid -> {
+                            boolean previousMultiThreadAcces = network.getVariantManager().isVariantMultiThreadAccessAllowed();
+                            network.getVariantManager().allowVariantMultiThreadAccess(true);
+                            CompletableFuture<Void>[] futures = new CompletableFuture[contingencies.size()];
+                            for (int i = 0; i < contingencies.size(); i++) {
+                                Contingency contingency = contingencies.get(i);
+
+                                // run one loadflow per contingency
+                                futures[i] = CompletableFuture
+                                        .supplyAsync(() -> {
+                                            try {
+                                                return queue.take();
+                                            } catch (InterruptedException e) {
+                                                Thread.currentThread().interrupt();
+                                                throw new UncheckedInterruptedException(e);
+                                            }
+                                        }, computationManager.getExecutor())
+                                        .thenCompose(workerId -> {
+                                            String postContStateId = hash + "_" + workerId;
+                                            return CompletableFuture
+                                                    .runAsync(() -> {
+                                                        network.getVariantManager().setWorkingVariant(postContStateId);
+                                                        // apply the contingency on the network
+                                                        contingency.toTask().modify(network, computationManager);
```

Comment:
agreed with @mathbagu : Here we should clone the variant used for pre contingency computation, in order to restore the "N" situation before applying the contingency.

Reply:
fixed